design/object-store-user.md

```diff
@@ -0,0 +1,131 @@
+
+#### Namespace
+When a storage class is specified the object store’s namespace is found in the storage class parameter. A typical object store user definition might look like this:
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: cephObjectStoreUser
+metadata:
+  name: my-user-1
+    namespace: dev-user
+spec:
+  storageClass: rook-ceph-object-store  # name of storage class
+  displayName: aUserDisplayName
+```
+Through the `storageClass` value we know the namespace and name of the object store.
+
+#### Secret OwnerReference
+Today, the generated key-pair secret’s owner reference is the id of the ceph cluster, which is also the same owner set for the CephObjectStoreUser. Thus, using the Kubernete’s [cascading deletes feature](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#foreground-cascading-deletion), the secret is not actually deleted until the ceph cluster is deleted. This enhancement proposes to set the Secret’s OwnerReference to the id of the CephObjectStoreUser that created the secret. So, when the CephObjectStoreUser is deleted, its secret is automatically deleted. A finalizer will also be added so that the the generated secret cannot be deleted until its ownerCephObjectStoreUser is first deleted.
+
+#### Secret Name
+See note [1] above. Essentially, if the object store name is embedded in the secret’s name then the app developer needs to know the name (and namespace) of the object store. This defeats the main benefit of adding a storage class property to the API.
+
+### Use Cases
+_As an app developer, I want to use `kubectl` to request the creation of an object user associated to an existing object store represented by a storage class_
+##### Steps
+  1. The Rook-Ceph operator, watching all namespaces, sees a new `CephObjectStoreUser` in the developer's namespace. The associated object store service is obtained via the referenced Storage Class’s  `spec.parameters.objectstore` field.
+  1. As done today, the operator creates a new S3 user in the object store. The credentials for this user are stored in a Secret now created in *same* namespace as the originating CephObjectStoreUser. Per the standard controller pattern, if an error occurs the operator will retry the create.
+  1. Per standard kubelet behavior, the app pod is blocked from running until it can access the Secret data.
+
+_As an app developer, I want to use `kubectl` to request the deletion of an object store user and expect the secret to also be deleted._
+##### Steps
+The kubectl API call to delete the object user triggers an automated cleanup sequence:
+  1. The CephObjectStoreUser is marked for deletion first and left in the foreground. Deletion is blocked (by Kubernetes) by an automated foreground finalizer which is removed once the secret is deleted.
```

Comment:
Is this referring to deleting the user secret from within Ceph? The way it's worded, it sounds like only K8s resources are being deleted, in which case we wouldn't need a finalizer.

Reply:
This refers to a user attempt to delete the `CephObjectStoreUser`. Yes, at this point in the design, only k8s resources are being deleted. IOW, the actual s3 key-pair associated with the bucket is not deleted, and the buckets containing these key-pairs (if any) are not modified. Delete here is strictly k8s housekeeping. I see a later phase where the `CephObjectStoreUser` resource has a _retention_ property added which can be set such that the key-pair and the s3 IAM user are physically deleted.